### DISCUSSION TITLE
How to disable edge runtime?

### DISCUSSION THREAD

**Discussion by MDOR** [vercel/next.js#38989] [Help] [answered]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

I need to create a custom middleware for authentication and log some information.

If I move the `midleware` to the root:
```js
export async function middleware(req) {
  try {
     logger.log(req)
  } catch(e) {
    console.error(e)
  } finally {
    req.next()
  }
} 

export const config = {
  runtime: 'nodejs',
};

```
I get errors because native APIs are not supported, but I need to use filesystem APIs.

I tried configuring this into the conf within `middleware` and `next.conf.js`.

Is there any way to get rid of the edge runtime? 

In my humble opinion is causing more harm than usefulness.

### How are you deploying your application? (if relevant)

mext start

### Describe the Bug

I need to avoid edge function usage on middleware.

### Expected Behavior

If properly configured, the edge function should not be used.

### Link to reproduction

Not applicable

### To Reproduce

Not applicable

**Comment by alexjs-dev** [comment]
Also having issues with edge runtime. Tried adding config to the middleware. Tried modifying next.config.. (Where I had to upgrade to react@18)

**Comment by audriuiv** [comment]
I can't disable too.

**Comment by lukebrandonfarrell** [comment]
Wow...

**Comment by lukebrandonfarrell** [comment]
You can use https://nextjs.org/docs/messages/edge-dynamic-code-evaluation `unstable_allowDynamic` to solve this issue. Export the `config` from your middleware file.

**Reply by cpetit636** [reply]
I have an issue with "next-logger". I'm unable to pass objects to the package and use its string formatting when using console.log with Edge. I need to completely disable Edge from the Middleware. Is it possible ?

**Comment by amplicity** [comment]
Yes -- there **needs** to be a way to disable edge runtime project-wide. Some of us have little need for edge, and it adds further complexity to simple implementations. I've been dealing with edge-related issues (via middleware, prisma, and now intermittent bugs that resolve on refresh) since I've started my next 13.4+ project.

It is absolutely unacceptable -- **there needs to be a global configuration to disable edge**, or it ends up creeping into places that you don't expect it, causing extra development time and ultimately eroding my patience.

**Reply by benjick** [reply]
Running NextJS in Docker and looking to disable edge project wide, can't believe there's no option for it

**Comment by amplicity** [comment]
FWIW OP -- you can write an api endpoint and have your middleware call that to workaround edge's issues/limitations and have full node access within the api endpoint.

**Reply by cpetit636** [reply]
all that for an access log ? no thanks. we now plan to replace next.js by another framework

**Reply by Fritze1602** [reply]
I just did that. All for simple "role" information from an existing non-Vercel PostgreSQL database. Prisma failed, pg failed too... . I couldn't think of any other option but to reach out to the "internet" and use a fetch call... . Hopefully, better options will be available soon.

**Comment by markspolakovs** [comment]
This prevents using database libraries like Prisma inside of middleware (https://github.com/prisma/prisma/issues/21310) - I can understand why you wouldn't want that when deploying to the edge, but it's a frustrating restriction when I'm self-hosting my application (my use case being checking user auth in middleware).

I've managed to work around this by putting all my authenticated routes in a route group and doing the auth check in its `layout.tsx`, but then I still need to manually add the check to route handlers rather than having the middleware handle this.

**Comment by msphn** [comment]
+1 same issue, edge is not viable and just another weird vendor lock feature. Allow us to turn that off for our self-hosted environments.

**Comment by leikoilja** [comment]
oh, this is absurd that one cannot turn off the edge runtime for the middleware.
i have been debugging production authentication issues (i am using supabase) for hours, while everything works perfectly locally, i came to realisation that the customer logic i have added to the middleware only tend to cause issues on the production because of enforced edge. :/

**Comment by hhunterx** [comment]
any news on this ?

**Comment by raphaeltm** [comment]
The fact that only the edge runtime is available is really disappointing. My expectation would be Node.js by default and opt-in to edge, given the limitations.

**Reply by e1himself** [reply]
> My expectation would be Node.js by default and opt-in to edge, given the limitations.

There is a benefit (for Vercel) in having middleware running in the Edge Runtime by default -- they want your Next.js app to be compatible with the Vercel infra, but not having a way to opt out of it is a big bummer.

**Reply by maiconsanson** [reply]
and a fee is charged for middleware executions on the edge... $0.65 per million.

**Comment by clayrisser** [comment]
Please add the ability to disable edge functions.

**Comment by laurens-mesure** [comment]
There's already an open issue for this. https://github.com/vercel/next.js/discussions/46722 unfortunately it's open since March 2023 and no word from the team.

**Comment by plepisnew** [comment]
*sigh*

**Comment by nikola-wd** [comment]
Deployed the standalone dir to Digital Ocean, because I want to manage my own infrastructure... I get this: 

```bash

/var/www/app/standalone/apps/web

node server.js

node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Cannot find module 'styled-jsx/package.json'
Require stack:
- /var/www/app/standalone/apps/web/node_modules/next/dist/server/require-hook.js
- /var/www/app/standalone/apps/web/node_modules/next/dist/server/next.js
- /var/www/app/standalone/apps/web/server.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at resolve (node:internal/modules/helpers:187:19)
    at Object.<anonymous> (/var/www/app/standalone/apps/web/node_modules/next/dist/server/require-hook.js:38:32)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/var/www/app/standalone/apps/web/node_modules/next/dist/server/next.js:25:1) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/var/www/app/standalone/apps/web/node_modules/next/dist/server/require-hook.js',
    '/var/www/app/standalone/apps/web/node_modules/next/dist/server/next.js',
    '/var/www/app/standalone/apps/web/server.js'
  ]
}

Node.js v20.11.1
```

**Reply by nikola-wd** [reply]
^ Just figured this out. Might help some people. The issue is that I am using PNPM, and it works with symlinks. In the GitHub (or whatever) pipeline, we need to make sure to preserve the symlinks in the artifacts when we upload to server. 

Besides the standalone dir, the static also needs to be uploaded (as it's stated in Next docs: https://nextjs.org/docs/pages/api-reference/next-config-js/output#automatically-copying-traced-files ).

**Comment by eero123456** [comment]
You could just patch the sandbox and run your own function there instead. Of course it's hacky but the whole thing is ridiculous. Running code on node in a js sandbox that emulates Edge...  And to make it more ironic, the sandbox already imports some node libraries ( buffer, events and such ) so you can add any library as well with some source edits. 

I had a project where to get db driver working I just patched the whole thing so there's a dummy middleware to "set things up" and then just run "my own" middleware after it. Fixing this officially should be trivial.

 try {
        // dummy blank next js middleware
        const result = await edgeFunction({
            request: {
                ...params.request,
                body: cloned && (0, _bodystreams.requestToBodyStream)(runtime.context, KUint8Array, cloned)
            }
        });
        for (const headerName of FORBIDDEN_HEADERS) {
            result.response.headers.delete(headerName);
        }
        // ***************************          
        const result_after_custom_middleware = await _customMiddleware(result);
        return result_after_custom_middleware;
        // ***************************************
    } finally {
        var _params_request_body1;
        await ((_params_request_body1 = params.request.body) == null ? void 0 : _params_request_body1.finalize());
    }

**Reply by benjick** [reply]
Hey, that's awesome. Could you provide some more details on how to actually do it? Would help me a lot to solve this

**Comment by mdrokz** [comment]
Is there still no update to this ? I want to check for user session using prisma in the middleware and i dont want to use the edge runtime for prisma and next.js

**Comment by BitBuf** [comment]
Kept encountering issues with this while implementing SSO with BoxyHQ. I've essentially been forced into using their cloud-hosted solution. Same deal with Prisma.

This needs to be a priority for at least Next 16. Or maybe even 15 to go with the other super obvious fixes.

**Comment by saullo** [comment]
The way the things is going, no middleware without edge on Nextjs 15.

**Comment by arielweinberger** [comment]
Please free us from edge runtime. It makes absolutely no sense that I deploy Next.js on Kubernetes with Docker, and I'm forced to be limited to edge runtime middleware, while the app is literally running on a Node.js 20.x Alpine image. Let us break the glass here, it's annoying.

**Comment by antoniopresto** [comment]
💩

**Comment by acoyfellow** [comment]
Is there literally no way to disable edge runtime? We can't have a "monorepo" next.js hosted on vercel?

**Comment by AghaWaleed** [comment]
pls

**Comment by duard** [comment]
the only way is changing framework

**Reply by acoyfellow** [reply]
Or provider... We migrated to Google Cloud Run.

**Comment by bbergeron-xlkey** [comment]
The edge runtime is a sad and regrettable constraint that I would've preferred to know about before migrating to Next.js. I was drawn to Next.js's popularity and was eager to work with it, but since I tried to implement a middleware and instrumentation, it now feels like it's working against me. It's grinding my gears and has brought the advancement of certain features to a halt, which I ironically thought would be faster to implement with Next.

I would very much like the folks at Vercel to hear us and give us the option to opt out of what should've been an opt-in solution from the start.

**Comment by clayrisser** [comment]
I don't need nextjs middleware anymore, because I don't need nextjs anymore.

**Comment by maxmandia** [comment]
update -- they're working on it

<img width="603" alt="image" src="https://github.com/user-attachments/assets/3ad6de18-db35-42f6-9a25-397437ceae21">

**Comment by maiconsanson** [comment]
For everyone who wanna know about the Node runtime and keep up with its updates: https://github.com/vercel/next.js/discussions/71727

**Comment by Abderrahmanlahmidi** [comment]
Configure Middleware to Use Node.js Runtime Set runtime: 'nodejs' in the middleware configuration to specify that the Node.js runtime should be used instead of the Edge runtime.

javascript
Copy code
// middleware.ts
import { NextRequest, NextResponse } from 'next/server';

export async function middleware(req: NextRequest) {
  try {
    // Your custom logging logic
    logger.log(req);
  } catch (e) {
    console.error(e);
  }
  return NextResponse.next(); // Finalizes the middleware without `req.next()`
}

export const config = {
  runtime: 'nodejs', // Enforce the Node.js runtime
};
Ensure next.config.js Is Configured Properly Double-check that your Next.js configuration does not force all middleware to Edge runtime. For example:

javascript
Copy code
// next.config.js
module.exports = {
  experimental: {
    middleware: {
      runtime: 'nodejs',
    },
  },
};
Alternative: Use an API Route for Node-Specific Logic If you still encounter issues, consider using an API route for tasks requiring filesystem access or other Node.js native APIs. The middleware can then redirect requests to this API route for the heavy-lifting server-side tasks.

Verify with Canary Release Since you’re on a canary release, confirm that this behavior persists by running npm update next@canary to ensure you’re on the latest canary version. Canary releases may have recent fixes or changes that improve middleware behavior.

Important Notes
req.next() should be replaced by NextResponse.next() in middleware.
Avoid using direct Node.js API imports in middleware to prevent unexpected Edge runtime behavior unless fully within the Node.js runtime.
This approach should let you avoid Edge limitations and use Node.js APIs smoothly in your middleware.

**Comment by leerob** [comment, marked as answer]
**We are working on allowing using the Node.js runtime for Middleware, and it will be included in an upcoming Next.js release.** We appreciate your feedback and patience on this issue.

We are also exploring a potential new API, called [Request Interceptors](https://github.com/vercel/next.js/pull/70961), that would address some of the feedback here for supporting nested, file-system based middleware. If this is something you are interested in, please comment on the proposal.

To prevent additional GitHub notifications and emails, we have a [**new discussion**](https://github.com/vercel/next.js/discussions/71727) where you can opt-into notifications as progress is made.

Thank you 🙏
